src/polymer/polymer2-mixin-scanner.ts

```diff
@@ -77,85 +77,77 @@ export class MixinVisitor implements Visitor {
-          summary: (summaryTag && summaryTag.description) || '',
-          privacy: getOrInferPrivacy(namespacedName, nodeJsDocs),
-          jsdoc: nodeJsDocs,
-          mixins: jsdoc.getMixinApplications(
-              this._document, node, nodeJsDocs, this.warnings, path.scope)
-        });
-        this._currentMixinNode = node;
-        this.mixins.push(this._currentMixin);
-      } else {
-        // Warn about a mixin whose name we can't infer.
-      }
+    if (!hasMixinFunctionDocTag(nodeJsDocs)) {
+      return;
+    }
+    const name = node.id.name;
+    const namespacedName =
+        name ? getNamespacedIdentifier(name, nodeJsDocs) : undefined;
+    if (namespacedName === undefined) {
+      // Warn about a mixin whose name we can't infer.
+      return;
     }
+    this.initializeMixin(node, namespacedName, nodeJsDocs, path.scope);
   }
 
   leaveFunctionDeclaration(
       node: babel.FunctionDeclaration, _parent: babel.Node) {
-    if (this._currentMixinNode === node) {
-      this._currentMixin = null;
-      this._currentMixinNode = null;
-      this._currentMixinFunction = null;
-    }
+    this.clearOnLeave(node);
   }
 
   enterVariableDeclaration(
       node: babel.VariableDeclaration, _parent: babel.Node, path: NodePath) {
-    const comment = esutil.getAttachedComment(node) || '';
+    const comment = esutil.getBestComment(path);
+    if (comment === undefined) {
+      return;
+    }
     const docs = jsdoc.parseJsdoc(comment);
-    const isMixin = hasMixinFunctionDocTag(docs);
-    const sourceRange = this._document.sourceRangeForNode(node)!;
-    const summaryTag = jsdoc.getTag(docs, 'summary');
-    if (isMixin) {
-      let mixin: ScannedPolymerElementMixin|undefined = undefined;
-      if (node.declarations.length === 1) {
-        const declaration = node.declarations[0];
-        const name = getIdentifierName(declaration.id);
-        if (name) {
-          const namespacedName = getNamespacedIdentifier(name, docs);
-          mixin = new ScannedPolymerElementMixin({
-            name: namespacedName,
-            sourceRange,
-            astNode: node,
-            description: docs.description,
-            summary: (summaryTag && summaryTag.description) || '',
-            privacy: getOrInferPrivacy(namespacedName, docs),
-            jsdoc: docs,
-            mixins: jsdoc.getMixinApplications(
-                this._document, declaration, docs, this.warnings, path.scope)
-          });
-        }
-      }
-      if (mixin) {
-        this._currentMixin = mixin;
-        this._currentMixinNode = node;
-        this.mixins.push(this._currentMixin);
-      } else {
-        // TODO(rictic); warn about being unable to determine mixin name.
-      }
+    if (!hasMixinFunctionDocTag(docs)) {
+      return;
+    }
+    if (node.declarations.length !== 1) {
+      return;
+    }
+    const declaration = node.declarations[0];
+    const name = getIdentifierName(declaration.id);
```

Comment:
worth noting that aside from getting the name, this is identical to `enterFunctionDeclaration`

Reply:
I figured out a way to unify this repetitive code.